firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/Datagrams.cpp

```diff
@@ -0,0 +1,1928 @@
+          "in the constructor") {
+        auto expected_payload = std::string("\x61\x6a\x1a\x6a\x29\xcf\x01\x81\xbe\x9d", 10);
+        REQUIRE(datagram.payload() == expected_payload);
+      }
+      THEN(
+          "The seq field of the body's header matches the value returned by the seq accessor "
+          "method") {
+        REQUIRE(first_body_output.operator[](0) == datagram.seq());
+      }
+      THEN(
+          "The length field of the body's header is equal to the size of the payload given in the "
+          "constructor") {
+        REQUIRE(first_body_output.operator[](1) == datagram.length());
+      }
+      THEN(
+          "The body's payload section correctly stores the paylaod as '0x61 0x6a 0x1a 0x6a 0x29 "
+          "0xcf 0x01 0x81 0xbe 0x9d'") {
+        for (size_t i = 2; i < 10; ++i) {
+          auto data = PF::Util::make_array<uint8_t>(
+              0x61, 0x6a, 0x1a, 0x6a, 0x29, 0xcf, 0x01, 0x81, 0xbe, 0x9d);
+          REQUIRE(first_body_output.operator[](i) == data[i - 2]);
+        }
+      }
+      THEN("The output buffer is as expected") {
+        auto expected_output = std::string("\x01\x0a\x61\x6a\x1a\x6a\x29\xcf\x01\x81\xbe\x9d", 12);
+        REQUIRE(first_body_output == expected_output);
+      }
+
+      // Parse
+      auto parse_status = parsed_datagram.parse(first_body_output);
+
+      THEN("The parse method reports ok status") { REQUIRE(parse_status == PF::IndexStatus::ok); }
+      THEN(
+          "After the parse method is called, The value returned by the seq accessor method is "
+          "equal to the sequence field of the input body's header") {
+        REQUIRE(parsed_datagram.seq() == 1);
+      }
+      THEN(
+          "The value returned by the length accessor method is equal to the length field of the "
+          "input body's header") {
+        REQUIRE(parsed_datagram.length() == 10);
+      }
+      THEN(
+          "The paylaod buffer returned by the paylaod accessor method is equal to the payload from "
+          "the body") {
+        auto expected_payload = std::string("\x61\x6a\x1a\x6a\x29\xcf\x01\x81\xbe\x9d", 10);
+        REQUIRE(parsed_datagram.payload() == expected_payload);
+      }
+      THEN(
+          "the payload given in the ParsedDatagram constructor is independent of the input buffer "
+          "body") {
+        output_buffer.push_back(0x02);
+        auto expected_payload = std::string("\x61\x6a\x1a\x6a\x29\xcf\x01\x81\xbe\x9d", 10);
+        REQUIRE(parsed_datagram.payload() == expected_payload);
+      }
+
+      // Write
+      // payload_output = parsed_datagram.payload()
+      auto payload_output = std::string("\x61\x6a\x1a\x6a\x29\xcf\x01\x81\xbe\x9d", 10);
+      TestDatagramProps::PayloadBuffer input_payload;
+      PF::Util::convert_string_to_byte_vector(payload_output, input_payload);
+      TestConstructedDatagram write_datagram(input_payload, 1);
```

Comment:
Was there a particular reason you're doing all this array copying and making a redundant hard-coded version of the data in line 738 (so that if we change line 738 but forget to change line 825, then the test continues to pass but it no longer behaves like a round-trip test even though the scenario calls it a round-trip test), instead of just doing TestConstructedDatagram write_datagram(parsed_payload, 1); or TestConstructedDatagram write_datagram(parsed_datagram.payload(), 1);?

Along these lines of maintainability, but a question of saving you work rather than about the risk of a test no longer doing what we say it does: is there a reasonable way you can define "\x61\x6a\x1a\x6a\x29\xcf\x01\x81\xbe\x9d" as a constant once, instead of hard-coded copy-pasting those bytes for all the REQUIREs?

Reply:
> instead of just doing TestConstructedDatagram write_datagram(parsed_payload, 1); or TestConstructedDatagram write_datagram(parsed_datagram.payload(), 1);

RIght, so as I had used a different payload there, instead of doing something like `parsed_datagram.payload()` i simply changed the payload bytes https://github.com/pez-globo/pufferfish-software/commit/06c3eed2ce934f6997dbb49c157c5a0521e80e3d#diff-d2aefa6418e0db5f2229e9f40480ace84b32ba511eaef5a49d967b3b0b869a4fR825
I'll fix this now.

> Along these lines of maintainability, but a question of potentially saving you work in the future rather than about the risk of a test no longer doing what we say it does: is there a reasonable way you can define "\x61\x6a\x1a\x6a\x29\xcf\x01\x81\xbe\x9d" as a constant once, instead of hard-coded copy-pasting those bytes for all the REQUIREs?

Will try to do that.